### DISCUSSION TITLE
Change mysql db location when installed with homebrew using Big Sur and external hard drive

### DISCUSSION THREAD

**Discussion by jamie-selesti** [Homebrew/discussions#2092] [Getting started]
Previously I had `/usr/local/var/mysql` symlinked to `/Volumes/External/mysql` meaning all my databases were stored on the external hard drive.

I have had to reformat my machine and upgrade to BigSur. If I try to set up the symlink as before I now get the following when I try to start MySQL

    brew services start mysql
    Bootstrap failed: 5: Input/output error
    Error: Failure while executing; `/bin/launchctl bootstrap gui/502 /Users/jamie/Library/LaunchAgents/homebrew.mxcl.mysql.plist` exited with 5. 

If I try to also change the `-datadir` in 

     /usr/local/Cellar/mysql/8.0.26/homebrew.mxcl.mysql.plist

to be 

    <string>--datadir=/Volumes/External/mysql</string>

I get the same error

    brew services start mysql
    Bootstrap failed: 5: Input/output error
    Error: Failure while executing; `/bin/launchctl bootstrap gui/502 /Users/jamie/Library/LaunchAgents/homebrew.mxcl.mysql.plist` exited with 5.

I have tried 

    launchctl unload /Users/jamie/Library/LaunchAgents/homebrew.mxcl.mysql.plist
    launchctl unload /Users/jamie/Library/LaunchAgents/homebrew.mxcl.mysql.plist

but that didn't work either. It's like it doesn't have the correct permissions.

Looking at the privacy settings you can see httpd which is also installed by brew is allowed to see "Removable volumes".

![Screenshot 2021-09-05 at 10 17 21](https://user-images.githubusercontent.com/1398649/132122104-a4ad366b-0398-4eef-adf5-a9f5e81e4d9e.png)

I can't add MySQL as the + symbol is grey out even though I have unlocked the panel

The external hard drive is located at `/Volumes/External/` and is APFS (Encrypted) volume.

Any advice would be great

**Comment by carlocab** [comment]
I think you should be able to produce the prompt to add permissions for it if you try to get `mysql` to access your external drive through normal use (i.e. not through `launchctl`). I don't use `mysql` so I don't know how exactly to do this, but I think this is what you want to go for.

Failing that, you can see if this might be helpful: https://apple.stackexchange.com/questions/376907/add-apps-to-files-and-folders-permissions

**Reply by jamie-selesti** [reply]
So this is strange...

If I stop the MySQL service via brew
`brew services stop mysql`

Then create a symblink to my external hard drive
`ln -s /Volumes/External/mysql /usr/local/var/mysql
`
Then manually start MySQL but launching /usr/local/Cellar/mysql/8.0.26/bin/mysqld it starts and works as expected.

If I then stop the MySQL services and and kill all connections using
`pkill mysql`

Then try to start MySQL via brew
`brew services start mysql`

It doesn't actually start

This means the symlink works when the service is not started by brew but doesnt work when it is started with brew.

Any ideas?

**Reply by jamie-selesti** [reply]
Right getting more to the issue now

If I edit /usr/local/Cellar/mysql/8.0.26/homebrew.mxcl.mysql.plist and add

```
	<key>StandardOutPath</key>
	<string>/tmp/homebrew.mxcl.mysql.stdout.log</string>
	<key>StandardErrorPath</key>
	<string>/tmp/homebrew.mxcl.mysql.stderr.log</string>
```

And then try 

`brew services start mysql`

In the log files I get


```
2021-09-14T16:37:55.6NZ mysqld_safe Starting mysqld daemon with databases from /usr/local/var/mysql
2021-09-14T16:37:55.6NZ mysqld_safe mysqld from pid file /usr/local/var/mysql/thor.local.pid ended
2021-09-14T16:38:05.6NZ mysqld_safe Logging to '/usr/local/var/mysql/thor.local.err'.

shell-init: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
job-working-directory: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
job-working-directory: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
job-working-directory: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
job-working-directory: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
pwd: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
pwd: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
chdir: error retrieving current directory: getcwd: cannot access parent directories: Operation not permitted
/usr/local/opt/mysql/bin/mysqld_safe: line 144: /usr/local/var/mysql/thor.local.err: Operation not permitted
/usr/local/opt/mysql/bin/mysqld_safe: line 144: /usr/local/var/mysql/thor.local.err: Operation not permitted
/usr/local/opt/mysql/bin/mysqld_safe: line 199: /usr/local/var/mysql/thor.local.err: Operation not permitted
/usr/local/opt/mysql/bin/mysqld_safe: line 144: /usr/local/var/mysql/thor.local.err: Operation not permitted
```
So its definitely a permissions issue just not sure how to fix it.

**Reply by jamie-selesti** [reply]
So figured it out 

https://www.symscape.com/node/1727

Select System Preferences->Security & Privacy->Full Disk Access
Click the lock to make changes
Click '+'
Press 'cmd + Shift + .' to show hidden files
select /bin/sh

**Comment by jamie-selesti** [comment]
Select System Preferences->Security & Privacy->Full Disk Access
Click the lock to make changes
Click '+'
Press 'cmd + Shift + .' to show hidden files
select /bin/sh
